source/trace_event.c

```diff
@@ -0,0 +1,353 @@
+}
+
+/* starts the message bus and initializes the JSON root, and the event array for storing metadata */
+int aws_trace_system_init(const char *filename, struct aws_allocator *allocator) {
+    if (allocator == NULL) {
+        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        return AWS_OP_ERR;
+    }
+
+    trace_event = aws_mem_calloc(allocator, 1, sizeof(struct aws_trace_event));
+    trace_event->allocator = allocator;
+
+    /* start the message bus */
+    /* Add option to select sync vs async? */
+    struct aws_bus_options options = {
+        .allocator = trace_event->allocator,
+        .policy = AWS_BUS_SYNC,
+    };
+
+    if (aws_bus_init(&(trace_event->bus), &options)) {
+        /* error occurs when bus->impl is not implemented */
+        aws_raise_error(AWS_ERROR_UNIMPLEMENTED);
+        goto error;
+    }
+
+    if (aws_bus_subscribe(&(trace_event->bus), 0, aws_trace_event_listener, NULL)) {
+        /* error occurs when bus's list of subscribers does not find or creater a spot for listener */
+        aws_raise_error(AWS_ERROR_LIST_EMPTY);
+        goto error;
+    }
+
+    trace_event->root = cJSON_CreateObject();
+
+    ERROR_CHECK(trace_event->root, AWS_ERROR_OOM);
+
+    trace_event->event_array = cJSON_CreateArray();
+    ERROR_CHECK(trace_event->event_array, AWS_ERROR_OOM);
+
+    cJSON_AddItemToObject(trace_event->root, "traceEvents", trace_event->event_array);
+
+    /* Set starting time for program */
+    if (aws_high_res_clock_get_ticks(&start_time)) {
+        aws_raise_error(AWS_ERROR_CLOCK_FAILURE);
+        goto error;
+    }
+
+    /* allocate memory and enable writing out if filename is not null */
+    if (filename != NULL) {
+        filename_out = aws_mem_acquire(trace_event->allocator, strlen(filename) + 1);
+        strcpy(filename_out, filename);
+    }
+
+    return AWS_OP_SUCCESS;
+
+error:
+    aws_trace_system_clean_up();
+    return AWS_OP_ERR;
+}
+
+int aws_trace_event(
+    const char *category,
+    const char *name,
+    char phase,
+    int value_1,
+    const char *value_name_1,
+    int value_2,
+    const char *value_name_2) {
+
+    /* set timestamps are in nanoseconds */
+    uint64_t timestamp;
+
+    if (aws_high_res_clock_get_ticks(&timestamp)) {
+        aws_raise_error(AWS_ERROR_CLOCK_FAILURE);
+        return AWS_OP_ERR;
+    }
+
+    timestamp -= start_time;
+    /* convert timestamps to tracing format of microseconds */
+    timestamp = aws_timestamp_convert(timestamp, AWS_TIMESTAMP_NANOS, AWS_TIMESTAMP_MICROS, 0);
+
+    /* get calling thread and process ids */
+    uint64_t thread_id = (uint64_t)aws_thread_current_thread_id();
+    int process_id = aws_get_pid();
+
+    struct aws_trace_event_metadata trace_event_data = {
+        .phase = phase,
+        .timestamp = timestamp,
+        .thread_id = thread_id,
+        .process_id = process_id,
+    };
+
+    strncpy(trace_event_data.name, name, 14);
+    strncpy(trace_event_data.category, category, 14);
```

Comment:
I think this will get cleaned up with some of the previous feedback, but try to avoid hardcoded constants for buffer sizes like this.  If someone changes the length of trace_event_data.category to something smaller, the 14 here becomes invalid, and it's not obvious to change.